### ISSUE TITLE
[FR] Folow best practices for REST API

### ISSUE THREAD

**Issue by mohammedi-haroune** [mlflow/mlflow#5128]
Thank you for submitting a feature request. **Before proceeding, please review MLflow's [Issue Policy for feature requests](https://www.github.com/mlflow/mlflow/blob/master/ISSUE_POLICY.md#feature-requests) and the [MLflow Contributing Guide](https://github.com/mlflow/mlflow/blob/master/CONTRIBUTING.rst)**.

**Please fill in this feature request template to ensure a timely and thorough response.**

## Willingness to contribute
The MLflow Community encourages new feature contributions. Would you or another member of your organization be willing to contribute an implementation of this feature (either as an MLflow Plugin or an enhancement to the MLflow code base)?

- [ ] Yes. I can contribute this feature independently.
- [x] Yes. I would be willing to contribute this feature with guidance from the MLflow community.
- [ ] No. I cannot contribute this feature at this time.

## Proposal Summary
Many tutorials suggests a set of best practices to follow when designing REST APIs:
- https://stackoverflow.blog/2020/03/02/best-practices-for-rest-api-design/
- https://docs.microsoft.com/en-us/azure/architecture/best-practices/api-design
- https://www.freecodecamp.org/news/rest-api-best-practices-rest-endpoint-design-examples/
- https://restfulapi.net/rest-api-design-tutorial-with-example/#model-uris

They have common practices:
- Use plural nouns and avoid verbs in endpoints
- Conform to HTTP semantics
- Use Nesting on Endpoints to Show Relationships
- Allow filtering, sorting, and pagination

MLFlow REST API violates most theses practices:
- it uses nouns in endpoint like `2.0/mlflow/runs/get`
- it  uses the HTTP request data to send resource id like `experiment_id` for [`UpdateExeperiment`](https://www.mlflow.org/docs/latest/rest-api.html#update-experiment)
- it uses `POST` to delete resources

This issue is about making the MLFlow REST API more standard by following the best practices.

If we do that, we'll have endpoints like:
- `/api/experiments`, `GET` to list the experiments `POST` to create a new experiment
- `/api/experiments/:id`, `GET` to get the experiment metadata, `PUT`/`PATCH` to update it, `DELETE` to delete it
- `/api/experiments/:id/runs`,  `GET` to list the runs of the experiment, `POST` to create a new run 
- `/api/runs/:id`, `GET` to get the run metadata, `PUT`/`PATCH` to update it, `DELETE` to delete it
- `/api/runs/:id/parameters`,  `GET` to get the parameters logged during that run, `POST` to log a new parameter
- `/api/runs/:id/parameters/:name`, `GET` to get the value of the parameter, `DELETE` to delete it
- `/api/runs/:id/metrics`, `GET` to get the metrics logged during that run, `POST` to log a new metric
- `/api/runs/:id/metric/:name`, `GET` to get the value of the metric, `DELETE` to delete it
- `/api/runs/:id/tags`, `GET` to get the tags logged during that run, `POST` to log new tag

## Motivation
- What is the use case for this feature?
- Why is this use case valuable to support for MLflow users in general?
- Why is this use case valuable to support for your project(s) or organization?
- Why is it currently difficult to achieve this use case? (please be as specific as possible about why related MLflow features and components are insufficient)

- This will make working with the REST API intuitive and easy to understand
- It'll make it possible to use Nginx's `request_auth` feature to implement authorization over MLFlow requests, it's not possible right now becuase Nginx `auth_request` erases the data before sending `/authorize` requests and MLFlow sends the `experiment_id/run_id` in the data of `POST` and `UPDATE` requests instead of the URL (POST /tracking/experiments/1), this makes it impossible to authorize such requests, see this comment: https://github.com/mlflow/mlflow/issues/761#issuecomment-981738147

### What component(s), interfaces, languages, and integrations does this feature affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [x] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interfaces
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Languages 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

## Details

(Use this section to include any additional information about the feature. If you have a proposal for how to implement this feature, please include it here. For implementation guidelines, please refer to the [Contributing Guide](https://github.com/mlflow/mlflow/blob/master/CONTRIBUTING.rst#contribution-guidelines).)

**Comment by amesar** [comment]
Good idea, but the reason the API is this way is that its based on Protobuf which doesn't lend itself to REST best practices. I doubt this will get accepted since Protobuf is all over the system and it would be major overhaul.
